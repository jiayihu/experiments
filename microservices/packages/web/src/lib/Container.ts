/**
 * Barebone dependency injection container
 */
export class Container<T extends Record<string, any>> {
  registry = new Map<keyof T, any>();

  bind<K extends keyof T>(name: K, value: T[K]) {
    this.registry.set(name, value);
  }

  get<K extends keyof T>(name: K): T[K] {
    return this.registry.get(name);
  }
}
