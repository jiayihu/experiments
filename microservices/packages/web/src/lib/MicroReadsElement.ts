import HyperHTMLElement from 'hyperhtml-element';

export class MicroReadElement<State = {}> extends HyperHTMLElement<State> {
  static define(name: string, options?: ElementDefinitionOptions) {
    HyperHTMLElement.define.call(this, name, options);

    const observedAttributes = this.observedAttributes || [];
    const proto = this.prototype;

    observedAttributes.forEach(name => {
      Object.defineProperty(proto, name, {
        configurable: true,
        get() {
          return this[`_${name}`] || super[name];
        },
        set(value) {
          if (typeof value === 'object' && value !== null) {
            this[`_${name}`] = value;
          } else {
            super[name] = value;
          }
        },
      });
    });
  }
}
