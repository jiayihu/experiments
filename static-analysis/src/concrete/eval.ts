import { Z, T, Name, Aexpr, Bexpr, Add, Mult, Sub, Eq, Le, Neg, And, Div } from '../syntax';
import { State } from './state';

type EvalAexpr = (aexpr: Aexpr) => (s: State) => Z;
type EvalBexpr = (bexpr: Bexpr) => (s: State) => T;

export const evalAexpr: EvalAexpr = (expr: Aexpr) => (s: State): Z => {
  switch (expr.type) {
    case 'Num':
      return expr.value;
    case 'Var':
      return s(expr.value);
    case 'Add':
      return evalAexpr(expr.aexpr1)(s) + evalAexpr(expr.aexpr2)(s);
    case 'Sub':
      return evalAexpr(expr.aexpr1)(s) - evalAexpr(expr.aexpr2)(s);
    case 'Mult':
      return evalAexpr(expr.aexpr1)(s) * evalAexpr(expr.aexpr2)(s);
    case 'Div':
      return evalAexpr(expr.aexpr1)(s) / evalAexpr(expr.aexpr2)(s);
  }
};

export const evalBexpr: EvalBexpr = (expr: Bexpr) => (s: State): T => {
  switch (expr.type) {
    case 'True':
      return expr.value;
    case 'False':
      return expr.value;
    case 'Eq':
      return evalAexpr(expr.aexpr1)(s) === evalAexpr(expr.aexpr2)(s);
    case 'Le':
      return evalAexpr(expr.aexpr1)(s) <= evalAexpr(expr.aexpr2)(s);
    case 'Neg':
      return !evalBexpr(expr.value)(s);
    case 'And': {
      return evalBexpr(expr.bexpr1)(s) && evalBexpr(expr.bexpr2)(s);
    }
  }
};

export const freeVariablesAexpr = (expr: Aexpr): Name[] => {
  switch (expr.type) {
    case 'Num':
      return [];
    case 'Var':
      return [expr.value];
    case 'Add':
    case 'Sub':
    case 'Mult':
    case 'Div': {
      const fvA1 = freeVariablesAexpr(expr.aexpr1);
      const fvA2 = freeVariablesAexpr(expr.aexpr2);

      return Array.from(new Set([...fvA1, ...fvA2]));
    }
  }
};

export const freeVariablesBexpr = (expr: Bexpr): Name[] => {
  switch (expr.type) {
    case 'True':
    case 'False':
      return [];
    case 'Eq':
    case 'Le': {
      const fvA1 = freeVariablesAexpr(expr.aexpr1);
      const fvA2 = freeVariablesAexpr(expr.aexpr2);

      return Array.from(new Set([...fvA1, ...fvA2]));
    }
    case 'Neg':
      return freeVariablesBexpr(expr.value);
    case 'And': {
      const fvB1 = freeVariablesBexpr(expr.bexpr1);
      const fvB2 = freeVariablesBexpr(expr.bexpr2);

      return Array.from(new Set([...fvB1, ...fvB2]));
    }
  }
};

export const substAexpr = (a: Aexpr) => (y: Name) => (a0: Aexpr): Aexpr => {
  switch (a.type) {
    case 'Num':
      return a;
    case 'Var':
      return a.value === y ? a0 : a;
    case 'Add':
      return new Add(substAexpr(a.aexpr1)(y)(a0), substAexpr(a.aexpr2)(y)(a0));
    case 'Sub':
      return new Sub(substAexpr(a.aexpr1)(y)(a0), substAexpr(a.aexpr2)(y)(a0));
    case 'Mult':
      return new Mult(substAexpr(a.aexpr1)(y)(a0), substAexpr(a.aexpr2)(y)(a0));
    case 'Div':
      return new Div(substAexpr(a.aexpr1)(y)(a0), substAexpr(a.aexpr2)(y)(a0));
  }
};

export const substBexpr = (b: Bexpr) => (y: Name) => (a0: Aexpr): Bexpr => {
  switch (b.type) {
    case 'True':
    case 'False':
      return b;
    case 'Eq':
      return new Eq(substAexpr(b.aexpr1)(y)(a0), substAexpr(b.aexpr2)(y)(a0));
    case 'Le':
      return new Le(substAexpr(b.aexpr1)(y)(a0), substAexpr(b.aexpr2)(y)(a0));
    case 'Neg':
      return new Neg(substBexpr(b.value)(y)(a0));
    case 'And':
      return new And(substBexpr(b.bexpr1)(y)(a0), substBexpr(b.bexpr2)(y)(a0));
  }
};

export const substState = (s: State) => (y: Name) => (v: Z): State => {
  return (x: Name) => (x === y ? v : s(x));
};
