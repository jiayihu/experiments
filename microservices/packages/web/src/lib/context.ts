import { Container } from './Container';
import hyperApp, { HyperHTMLApplication } from 'hyperhtml-app';
import { Store } from 'redux';
import { store } from '../redux/store';

type Context = {
  router: HyperHTMLApplication;
  store: Store;
};

const context = new Container<Context>();

context.bind('router', new hyperApp());
context.bind('store', store);

export { context };
