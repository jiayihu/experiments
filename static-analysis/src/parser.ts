import { span, split, isEmpty } from 'fp-ts/lib/Array';
import { Char } from 'newtype-ts/lib/Char';
import { iso } from 'newtype-ts';
import {
  twoCharOps,
  keywords,
  Var,
  Num,
  Aexpr,
  Add,
  Mult,
  Sub,
  True,
  False,
  Eq,
  Le,
  Neg,
  And,
  Bexpr,
  Stm,
  Ass,
  Skip,
  Comp,
  If,
  While,
} from './language';
import { Option, some, none } from 'fp-ts/lib/Option';
import { pair } from './utils';

const isoChar = iso<Char>();
export const strToChars = (xs: string): Char[] => Array.from(xs).map(isoChar.wrap);
export const charsToStr = (chars: Char[]): string => chars.map(isoChar.unwrap).join('');

type Token = Char[];

const isAlpha = (c: Char) => /^[A-Za-z_]$/.test(isoChar.unwrap(c));
const isAlphaNum = (c: Char) => /^\w$/.test(isoChar.unwrap(c));
const isDigit = (c: Char) => /^\d$/.test(isoChar.unwrap(c));
const isSpace = (c: Char) => /^\s$/.test(isoChar.unwrap(c));

export const tokenizer = (input: Char[]): Token[] => {
  if (!input.length) return [];

  const [c, ...cs] = input;
  const spanToken = (pred: (char: Char) => boolean): { token: Token; rest: Char[] } => {
    const { init, rest } = span(input, pred);
    return { token: init, rest };
  };

  if (isSpace(c)) return tokenizer(cs);
  if (isDigit(c)) {
    const { token, rest } = spanToken(isDigit);
    return [token, ...tokenizer(rest)];
  }
  if (isAlpha(c)) {
    const { token, rest } = spanToken(isAlphaNum);
    return [token, ...tokenizer(rest)];
  }
  if (input.length >= 2) {
    const [twoChars, rest] = split(2, input);
    if (twoCharOps.includes(charsToStr(twoChars))) {
      return [twoChars, ...tokenizer(rest)];
    }
  }

  return [[c], ...tokenizer(cs)];
};

/**
 * Parser Monad, basically a State monad implementation where the state is the
 * list of tokens but the computation can fail.
 * {@link https://github.com/gcanti/fp-ts/blob/master/src/State.ts#L19-L65}
 */
export class Parser<A> {
  constructor(readonly parse: (tokens: Token[]) => Option<[A, Token[]]>) {}

  static of = <A>(a: A): Parser<A> => {
    return new Parser(tokens => some(pair([a, tokens])));
  };

  map<B>(f: (a: A) => B): Parser<B> {
    return new Parser(tokens => {
      return this.parse(tokens).map(([v, restTokens]) => {
        return pair([f(v), restTokens]);
      });
    });
  }

  ap<B>(fab: Parser<(a: A) => B>): Parser<B> {
    return fab.chain(f => this.map(f));
  }

  ap_<B, C>(this: Parser<(b: B) => C>, fb: Parser<B>): Parser<C> {
    return fb.ap(this);
  }

  chain<B>(f: (a: A) => Parser<B>): Parser<B> {
    return new Parser(tokens => {
      return this.parse(tokens).chain(([v, restTokens]) => {
        return f(v).parse(restTokens);
      });
    });
  }

  static empty = <A>(): Parser<A> => new Parser(_ => none);

  alt(fa: Parser<A>): Parser<A> {
    return new Parser(tokens => {
      return this.parse(tokens).orElse(() => fa.parse(tokens));
    });
  }
}

const item: Parser<Token> = new Parser(tokens => {
  if (isEmpty(tokens)) return none;

  const [x, ...xs] = tokens;
  return some(pair([x, xs]));
});

const pOneOrMoreWithSep = <A, B>(pa: Parser<A>) => (pb: Parser<B>): Parser<A[]> => {
  return pa.chain(a =>
    pb
      .chain(() => pOneOrMoreWithSep(pa)(pb))
      .chain(as => Parser.of([a, ...as]))
      .alt(Parser.of([a])),
  );
};

const pSat = (pred: (token: Token) => boolean): Parser<Token> => {
  return item.chain(token => (pred(token) ? Parser.of(token) : Parser.empty()));
};

const pLit = (c: Token): Parser<Token> => pSat(token => token.join('') === c.join(''));

/**
 * Arithmetic expression parsers, it handles right-associativity and Mult precedence.
 * Based on the following grammar:
 *
 * aexpr ::= term (+ aexpr | ε) | term (- aexpr | ε)
 * term ::= factor (* aexpr | ε)
 * factor ::= (aexpr) | basis
 * basic ::= Var | Num
 */

const pVar: Parser<Var> = pSat(
  token => !keywords.includes(charsToStr(token)) && isAlpha(token[0]),
).map(token => new Var(charsToStr(token)));

const pNum: Parser<Num> = pSat(token => token.every(isDigit)).map(
  token => new Num(Number(token.join(''))),
);

const pBasisAexpr: Parser<Aexpr> = (pVar as Parser<Aexpr>).alt(pNum);

const pFactor: Parser<Aexpr> = pLit(strToChars('('))
  .chain(() => pAexpr)
  .chain(a => pLit(strToChars(')')).map(() => a))
  .alt(pBasisAexpr);

const pTerm: Parser<Aexpr> = pFactor.chain(a1 =>
  pLit(strToChars('*'))
    .chain(() => pTerm)
    .map(a2 => new Mult(a1, a2) as Aexpr)
    .alt(Parser.of(a1)),
);

const pAdd: Parser<Aexpr> = pTerm.chain(a1 =>
  pLit(strToChars('+'))
    .chain(() => pAexpr)
    .map(a2 => new Add(a1, a2)),
);

const pSub: Parser<Aexpr> = pTerm.chain(a1 =>
  pLit(strToChars('-'))
    .chain(() => pAexpr)
    .map(a2 => new Sub(a1, a2)),
);

export const pAexpr: Parser<Aexpr> = pAdd.alt(pSub).alt(pTerm);

/**
 * Boolean expression parsers
 */

const pTrue: Parser<True> = pSat(token => charsToStr(token) === 'true').map(() => new True(true));

const pFalse: Parser<False> = pSat(token => charsToStr(token) === 'false').map(
  () => new False(false),
);

const pEq: Parser<Eq> = pAexpr.chain(a1 =>
  pLit(strToChars('='))
    .chain(() => pAexpr)
    .map(a2 => new Eq(a1, a2)),
);

const pLe: Parser<Le> = pAexpr.chain(a1 =>
  pLit(strToChars('<'))
    .chain(() => pLit(strToChars('=')))
    .chain(() => pAexpr)
    .map(a2 => new Le(a1, a2)),
);

const pBasisBexpr: Parser<Bexpr> = (pTrue as Parser<Bexpr>)
  .alt(pFalse)
  .alt(pEq)
  .alt(pLe);

const pNeg: Parser<Neg> = pLit(strToChars('¬'))
  .chain(() => pBexpr)
  .map(b => new Neg(b));

const pAnd: Parser<And> = pBasisBexpr.chain(a1 =>
  pLit(strToChars('∧'))
    .chain(() => pBasisBexpr)
    .map(a2 => new And(a1, a2)),
);

const pParBexpr: Parser<Bexpr> = pLit(strToChars('('))
  .chain(() => pBexpr)
  .chain(a => pLit(strToChars(')')).map(() => a));

export const pBexpr: Parser<Bexpr> = (pNeg as Parser<Bexpr>)
  .alt(pAnd)
  .alt(pBasisBexpr)
  .alt(pParBexpr);

/**
 * Statement parsers
 */

const pAss: Parser<Ass> = pSat(token => isAlpha(token[0])).chain(name =>
  pLit(strToChars(':='))
    .chain(() => pAexpr)
    .map(a2 => new Ass(name.join(''), a2)),
);

const pSkip: Parser<Skip> = pLit(strToChars('skip')).map(() => new Skip());

const pIf: Parser<If> = pLit(strToChars('if'))
  .chain(() => pBexpr)
  .chain(bexpr =>
    pLit(strToChars('then'))
      .chain(() => pStm)
      .chain(stm1 =>
        pLit(strToChars('else'))
          .chain(() => pStm)
          .map(stm2 => new If(bexpr, stm1, stm2)),
      ),
  );

const pWhile: Parser<While> = pLit(strToChars('while'))
  .chain(() => pBexpr)
  .chain(bexpr =>
    pLit(strToChars('do'))
      .chain(() => pProg)
      .map(stm => new While(bexpr, stm)),
  );

const pStm: Parser<Stm> = (pAss as Parser<Stm>)
  .alt(pSkip)
  .alt(pIf)
  .alt(pWhile);

const pParStm: Parser<Stm> = pLit(strToChars('('))
  .chain(() => pProg)
  .chain(stm => pLit(strToChars(')')).map(() => stm));

export const pProg: Parser<Stm> = pOneOrMoreWithSep(pStm.alt(pParStm))(pLit(strToChars(';'))).map(
  stms =>
    stms.reduce<Stm>((comp, stm) => {
      // Cannot avoid putting Skip as initial value so if the program is only one stm avoid Comp
      return comp.type === 'Skip' ? stm : new Comp(comp, stm);
    }, new Skip()),
);
