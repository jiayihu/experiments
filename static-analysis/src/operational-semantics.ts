import { Stm, State, If, Comp, While, Skip } from './syntax';
import { substState, evalAexpr, evalBexpr } from './eval';

export class InterConfig {
  type: 'InterConfig' = 'InterConfig';
  constructor(readonly stm: Stm, readonly state: State) {}
}

export class FinalConfig {
  type: 'FinalConfig' = 'FinalConfig';
  constructor(readonly state: State) {}
}

export type Config = InterConfig | FinalConfig;

const isFinal = (config: Config): config is FinalConfig => {
  return config.type === 'FinalConfig';
};

export const transition = (config: Config): Config => {
  if (isFinal(config)) return config;

  const { stm, state } = config;

  switch (stm.type) {
    case 'Ass': {
      const { name, aexpr } = stm;
      return new FinalConfig(substState(state)(name)(evalAexpr(aexpr)(state)));
    }
    case 'Skip': {
      return new FinalConfig(state);
    }
    case 'Comp': {
      const { stm1, stm2 } = stm;
      const config1 = transition(new InterConfig(stm1, state));

      return isFinal(config1)
        ? new InterConfig(stm2, config1.state)
        : new InterConfig(new Comp(config1.stm, stm2), config1.state);
    }
    case 'If': {
      const { bexpr, stm1, stm2 } = stm;

      return evalBexpr(bexpr)(state) ? new InterConfig(stm1, state) : new InterConfig(stm2, state);
    }
    case 'While': {
      const { bexpr, stm: whileStm } = stm;

      return new InterConfig(
        new If(bexpr, new Comp(whileStm, new While(bexpr, whileStm)), new Skip()),
        state,
      );
    }
  }
};

export const deriveSeq = (config: Config): Config[] => {
  if (isFinal(config)) return [config];

  return [config, ...deriveSeq(transition(config))];
};

export const semantic = (stm: Stm) => (state: State): State => {
  const derivation = deriveSeq(new InterConfig(stm, state));
  const last = derivation[derivation.length - 1];

  return last.state;
};
