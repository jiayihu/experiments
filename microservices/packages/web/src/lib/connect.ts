import { Store, Unsubscribe, Action } from 'redux';
import { context } from './context';
import HyperHTMLElement from 'hyperhtml-element';

type Constructor<T> = new (...args: any[]) => T;

interface CustomElement {
  connectedCallback?(): void;
  disconnectedCallback?(): void;
  readonly isConnected: boolean;
}

/**
 * Allows to connect a Web Component to Redux
 * Implementation is taken from @link https://github.com/Polymer/pwa-helpers
 */
export const connect = <S>(store: Store<S>) => <T extends Constructor<CustomElement>>(
  baseElement: T,
) =>
  class extends baseElement {
    _storeUnsubscribe!: Unsubscribe;

    connectedCallback() {
      if (super.connectedCallback) {
        super.connectedCallback();
      }

      this._storeUnsubscribe = store.subscribe(() => this.stateChanged(store.getState()));
      this.stateChanged(store.getState());
    }

    disconnectedCallback() {
      this._storeUnsubscribe();

      if (super.disconnectedCallback) {
        super.disconnectedCallback();
      }
    }

    /**
     * The `stateChanged(state)` method will be called when the state is updated.
     */
    stateChanged(_state: S) {}

    dispatch(action: Action) {
      store.dispatch(action);
    }
  };

export const ConnectedHyperElement = connect(context.get('store'))(HyperHTMLElement);
