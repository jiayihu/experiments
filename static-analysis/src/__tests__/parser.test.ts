import {
  Add,
  Var,
  Num,
  Eq,
  Ass,
  Neg,
  Sub,
  Mult,
  And,
  Div,
  Le,
  True,
  False,
  Skip,
  If,
  While,
  initialState,
  Comp,
} from '../syntax';
import { strToChars, tokenizer, charsToStr, pAexpr, pBexpr, pProg } from '../parser';
import { some, option } from 'fp-ts/lib/Option';
import { array } from 'fp-ts/lib/Array';
import { evalBexpr, evalAexpr } from '../eval';
import { semantic } from '../operational-semantics';

describe('tokenizer', () => {
  it('should split in tokens', () => {
    const input = strToChars('x + 1');
    const tokens = tokenizer(input).map(charsToStr);

    expect(tokens).toEqual(['x', '+', '1']);
  });

  it('should split in tokens', () => {
    const input = strToChars('x + y * x');
    const tokens = tokenizer(input).map(charsToStr);

    expect(tokens).toEqual(['x', '+', 'y', '*', 'x']);
  });

  it('should handle two-char operators', () => {
    const input = strToChars('x <= 2');
    const tokens = tokenizer(input).map(charsToStr);

    expect(tokens).toEqual(['x', '<=', '2']);
  });
});

describe('parser', () => {
  it('should parse simple Add expression', () => {
    const input = strToChars('x + 1');
    const tokens = tokenizer(input);

    expect(pAexpr.parse(tokens)).toEqual(some([new Add(new Var('x'), new Num(1)), []]));
  });

  it('should parse simple Sub expression', () => {
    const input = strToChars('x - 1');
    const tokens = tokenizer(input);

    expect(pAexpr.parse(tokens)).toEqual(some([new Sub(new Var('x'), new Num(1)), []]));
  });

  it('should give precedence to Mult operator', () => {
    const input = strToChars('2 * 3 + 4');
    const tokens = tokenizer(input);

    expect(pAexpr.parse(tokens)).toEqual(
      some([new Add(new Mult(new Num(2), new Num(3)), new Num(4)), []]),
    );
  });

  it('should give precedence to Mult operator', () => {
    const input = strToChars('2 + 3 * 4');
    const tokens = tokenizer(input);

    expect(pAexpr.parse(tokens)).toEqual(
      some([new Add(new Num(2), new Mult(new Num(3), new Num(4))), []]),
    );
  });

  it('should give precedence to parenthesis', () => {
    const input = strToChars('2 * (3 + 4)');
    const tokens = tokenizer(input);

    expect(pAexpr.parse(tokens)).toEqual(
      some([new Mult(new Num(2), new Add(new Num(3), new Num(4))), []]),
    );
  });

  it('should associate Aexpr to right', () => {
    const input = strToChars('2 + 3 + 4');
    const tokens = tokenizer(input);

    expect(pAexpr.parse(tokens)).toEqual(
      some([new Add(new Num(2), new Add(new Num(3), new Num(4))), []]),
    );
  });

  it('should parse the battery of Aexpr tests', () => {
    const input = ['x+2*y', 'x*(2*y)', '(x*2)*y', '(x/3)*2', 'x/3*2'];
    const tokens = input.map(aexpr => tokenizer(strToChars(aexpr)));
    const syntaxes = tokens.map(ts => pAexpr.parse(ts));
    const result = array.sequence(option)(syntaxes);

    expect(result).toEqual(
      some([
        [new Add(new Var('x'), new Mult(new Num(2), new Var('y'))), []],
        [new Mult(new Var('x'), new Mult(new Num(2), new Var('y'))), []],
        [new Mult(new Mult(new Var('x'), new Num(2)), new Var('y')), []],
        [new Mult(new Div(new Var('x'), new Num(3)), new Num(2)), []],
        [new Div(new Var('x'), new Mult(new Num(3), new Num(2))), []],
      ]),
    );
  });

  it('should handle syntactic sugar', () => {
    const input = ['-1 + 2', '-(1 + 2)', '3 - 1'];
    const tokens = input.map(bexpr => tokenizer(strToChars(bexpr)));
    const syntaxes = tokens
      .map(ts => pAexpr.parse(ts))
      .map(syntax => syntax.map(([aexpr]) => evalAexpr(aexpr)(() => 0)));
    const result = array.sequence(option)(syntaxes);
    const expected = some([1, -3, 2]);

    expect(result).toEqual(expected);
  });

  it('should parse basis boolean expression', () => {
    const input = strToChars('3 = 2');
    const tokens = tokenizer(input);

    expect(pBexpr.parse(tokens)).toEqual(some([new Eq(new Num(3), new Num(2)), []]));
  });

  it('should parse neg boolean expression', () => {
    const input = strToChars('!(x = 1)');
    const tokens = tokenizer(input);

    expect(pBexpr.parse(tokens)).toEqual(some([new Neg(new Eq(new Var('x'), new Num(1))), []]));
  });

  it('should give precedence to Neg operator', () => {
    const input = strToChars('!(x = 1) & (x = 2)');
    const tokens = tokenizer(input);

    expect(pBexpr.parse(tokens)).toEqual(
      some([
        new And(new Neg(new Eq(new Var('x'), new Num(1))), new Eq(new Var('x'), new Num(2))),
        [],
      ]),
    );
  });

  it('should associate Bexpr to right', () => {
    const input = strToChars('x = 3 & x = 4 & x = 5');
    const tokens = tokenizer(input);

    expect(pBexpr.parse(tokens)).toEqual(
      some([
        new And(
          new Eq(new Var('x'), new Num(3)),
          new And(new Eq(new Var('x'), new Num(4)), new Eq(new Var('x'), new Num(5))),
        ),
        [],
      ]),
    );
  });

  it('should parse the battery of Bexpr tests', () => {
    const input = [
      '(4+(3)) <= 5',
      '!true & (8 <= 2)',
      '!true & false',
      '!(true & false)',
      '!((4 = 8) & false)',
    ];
    const tokens = input.map(aexpr => tokenizer(strToChars(aexpr)));
    const syntaxes = tokens.map(ts => pBexpr.parse(ts));
    const result = array.sequence(option)(syntaxes);

    expect(result).toEqual(
      some([
        [new Le(new Add(new Num(4), new Num(3)), new Num(5)), []],
        [new And(new Neg(new True()), new Le(new Num(8), new Num(2))), []],
        [new And(new Neg(new True()), new False()), []],
        [new Neg(new And(new True(), new False())), []],
        [new Neg(new And(new Eq(new Num(4), new Num(8)), new False())), []],
      ]),
    );
  });

  it('should handle syntactic sugar', () => {
    const input = [
      '5 != 5',
      '5 != 4',
      '5 < 4',
      '5 < 6',
      '5 < 5',
      '6 > 7',
      '6 > 5',
      '6 > 6',
      '6 >= 5',
      '6 >= 6',
      '6 >= 7',
      '5 <= 3 | 5 <= 3',
      '5 <= 3 | 3 <= 5',
      '3 <= 5 | 3 <= 5',
    ];
    const tokens = input.map(bexpr => tokenizer(strToChars(bexpr)));
    const syntaxes = tokens
      .map(ts => pBexpr.parse(ts))
      .map(syntax => syntax.map(([bexpr]) => evalBexpr(bexpr)(() => 0)));
    const result = array.sequence(option)(syntaxes);
    const expected = some([
      false,
      true,
      false,
      true,
      false,
      false,
      true,
      false,
      true,
      true,
      false,
      false,
      true,
      true,
    ]);

    expect(result).toEqual(expected);
  });

  it('should parse simple statement', () => {
    const input = strToChars('y := 1');
    const tokens = tokenizer(input);

    expect(pProg.parse(tokens)).toEqual(some([new Ass('y', new Num(1)), []]));
  });

  it('should parse the battery of Stm tests', () => {
    const input = [
      'skip',
      'x := 3',
      'if 3 <= 33 then skip else x := 1',
      'while true do skip',
      '(skip)',
    ];
    const tokens = input.map(aexpr => tokenizer(strToChars(aexpr)));
    const syntaxes = tokens.map(ts => pProg.parse(ts));
    const result = array.sequence(option)(syntaxes);

    expect(result).toEqual(
      some([
        [new Skip(), []],
        [new Ass('x', new Num(3)), []],
        [new If(new Le(new Num(3), new Num(33)), new Skip(), new Ass('x', new Num(1))), []],
        [new While(new True(), new Skip()), []],
        [new Skip(), []],
      ]),
    );
  });

  it('should parse composed if', () => {
    const input = strToChars('if 1 = 1 then skip else (x := 1; y := 2)');
    const tokens = tokenizer(input);

    expect(pProg.parse(tokens)).toEqual(
      some([
        new If(
          new Eq(new Num(1), new Num(1)),
          new Skip(),
          new Comp(new Ass('x', new Num(1)), new Ass('y', new Num(2))),
        ),
        [],
      ]),
    );
  });

  it('should parse factorial statement', () => {
    const input = strToChars('y := 1; while !(x = 1) do (y := y * x; x := x - 1)');
    const tokens = tokenizer(input);

    expect(pProg.parse(tokens)).toMatchSnapshot();
  });

  it('should parse composed while', () => {
    const input = strToChars('y := 1; while !(x = 1) do (y := y * x; x := x - 1); z := 1');
    const tokens = tokenizer(input);

    expect(pProg.parse(tokens)).toMatchSnapshot();
  });

  it('should handle for syntactic sugar', () => {
    const input = strToChars('for x := 1 to 10 do y := y + x');
    const tokens = tokenizer(input);
    const syntax = pProg.parse(tokens);
    const state = initialState({ y: 1 });
    const result = syntax.map(([stm]) => semantic(stm)(state)).map(s => s('y'));

    expect(result).toEqual(some(56));
  });

  it('should handle repeat until syntactic sugar', () => {
    const input = strToChars('repeat x := x + 1 until x = 10');
    const tokens = tokenizer(input);
    const syntax = pProg.parse(tokens);
    const state = initialState({ x: 0 });
    const result = syntax.map(([stm]) => semantic(stm)(state)).map(s => s('x'));

    expect(result).toEqual(some(10));
  });

  it('should execute repeat until S at least once', () => {
    const input = strToChars('repeat x := x + 1 until 1 = 1');
    const tokens = tokenizer(input);
    const syntax = pProg.parse(tokens);
    const state = initialState({ x: 0 });
    const result = syntax.map(([stm]) => semantic(stm)(state)).map(s => s('x'));

    expect(result).toEqual(some(1));
  });
});
