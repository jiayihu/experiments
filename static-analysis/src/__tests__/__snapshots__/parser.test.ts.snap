// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse composed while 1`] = `
Some {
  "_tag": "Some",
  "value": Array [
    Comp {
      "stm1": Ass {
        "aexpr": Num {
          "type": "Num",
          "value": 1,
        },
        "name": "y",
        "type": "Ass",
      },
      "stm2": Comp {
        "stm1": While {
          "bexpr": Neg {
            "type": "Neg",
            "value": Eq {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Eq",
            },
          },
          "stm": Comp {
            "stm1": Ass {
              "aexpr": Mult {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "y",
                },
                "aexpr2": Var {
                  "type": "Var",
                  "value": "x",
                },
                "type": "Mult",
              },
              "name": "y",
              "type": "Ass",
            },
            "stm2": Ass {
              "aexpr": Sub {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "x",
                },
                "aexpr2": Num {
                  "type": "Num",
                  "value": 1,
                },
                "type": "Sub",
              },
              "name": "x",
              "type": "Ass",
            },
            "type": "Comp",
          },
          "type": "While",
        },
        "stm2": Ass {
          "aexpr": Num {
            "type": "Num",
            "value": 1,
          },
          "name": "z",
          "type": "Ass",
        },
        "type": "Comp",
      },
      "type": "Comp",
    },
    Array [],
  ],
}
`;

exports[`parser should parse factorial statement 1`] = `
Some {
  "_tag": "Some",
  "value": Array [
    Comp {
      "stm1": Ass {
        "aexpr": Num {
          "type": "Num",
          "value": 1,
        },
        "name": "y",
        "type": "Ass",
      },
      "stm2": While {
        "bexpr": Neg {
          "type": "Neg",
          "value": Eq {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Eq",
          },
        },
        "stm": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "type": "While",
      },
      "type": "Comp",
    },
    Array [],
  ],
}
`;
